core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceCreateSpec.scala

```diff
@@ -0,0 +1,849 @@
+          testContext
+        ),
+        Duration.Inf
+      )
+    }
+
+    e.errorReport.statusCode shouldBe Some(StatusCodes.InternalServerError)
+    verify(workspaceRepository).deleteWorkspace(workspace.toWorkspaceName)
+    verify(wsmDAO).deleteWorkspaceV2(ArgumentMatchers.eq(workspaceId), any(), any())
+    verify(wsmDAO).getDeleteWorkspaceV2Result(workspaceId, deleteJobId.toString, testContext)
+  }
+
+  it should "fail and rollback workspace creation on async workspace creation job failure in WSM" in {
+    val workspace = defaultWorkspace
+    val workspaceRepository = mock[WorkspaceRepository]
+    when(
+      workspaceRepository.createMCWorkspace(
+        ArgumentMatchers.eq(workspaceId),
+        ArgumentMatchers.eq(WorkspaceName(namespace, name)),
+        ArgumentMatchers.any(),
+        ArgumentMatchers.eq(testContext),
+        ArgumentMatchers.any()
+      )(ArgumentMatchers.any())
+    ).thenReturn(Future(workspace))
+    when(workspaceRepository.deleteWorkspace(workspace.toWorkspaceName)).thenReturn(Future(true))
+    val samDAO = mock[SamDAO]
+    when(samDAO.userHasAction(any(), any(), any(), any())).thenReturn(Future(true))
+    val wsmDAO = mock[WorkspaceManagerDAO]
+    val createJobId = UUID.randomUUID()
+    when(wsmDAO.createWorkspaceWithSpendProfile(any(), any(), any(), any(), any(), any(), any(), any()))
+      .thenReturn(new CreateWorkspaceV2Result().jobReport(new JobReport().id(createJobId.toString)))
+    // This is the primary failure we're testing
+    when(wsmDAO.getCreateWorkspaceResult(createJobId.toString, testContext))
+      .thenReturn(new CreateWorkspaceV2Result().jobReport(new JobReport().status(StatusEnum.FAILED)))
+    val deleteJobId = UUID.randomUUID()
+    when(wsmDAO.deleteWorkspaceV2(ArgumentMatchers.eq(workspaceId), any(), any()))
+      .thenReturn(new JobResult().jobReport(new JobReport().id(deleteJobId.toString)))
+    when(wsmDAO.getDeleteWorkspaceV2Result(workspaceId, deleteJobId.toString, testContext))
+      .thenReturn(new JobResult().jobReport(new JobReport().id(deleteJobId.toString).status(StatusEnum.SUCCEEDED)))
+
+    val service = new MultiCloudWorkspaceService(
+      testContext,
+      wsmDAO,
+      mock[BillingProfileManagerDAO],
+      samDAO,
+      MultiCloudWorkspaceConfig(MultiCloudWorkspaceManagerConfig("app", 60 seconds, 120 seconds), mock[AzureConfig]),
+      mock[LeonardoDAO],
+      "MultiCloudWorkspaceService-test",
+      mock[WorkspaceManagerResourceMonitorRecordDao],
+      workspaceRepository,
+      mock[BillingRepository]
+    )
+
+    val e = intercept[RawlsExceptionWithErrorReport] {
+      Await.result(
+        service.createMultiCloudWorkspaceInt(
+          WorkspaceRequest(namespace, name, Map.empty),
+          workspaceId,
+          new ProfileModel().id(UUID.randomUUID()),
+          testContext
+        ),
+        Duration.Inf
+      )
+    }
+
+    e.errorReport.statusCode shouldBe Some(StatusCodes.InternalServerError)
```

Comment:
Nit / style difference: when there are several things that all have to be checked as a block, in multiple test cases, I think having a helper method like `verifyWorkspaceCreationRollback` is actually quite nice.

Reply:
I find that those tend to get hard to maintain quickly, and often end up lying about what they're doing.
I agree that it _can_ be worth it if you have a lot of the same things that have the same verifications many times; I think I probably just have a higher threshold.